pandas/core/frame.py

```diff
@@ -3952,6 +3953,25 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         if not isinstance(keys, list):
             keys = [keys]
 
+        missing = []
+        for col in keys:
+            if (is_scalar(col) or isinstance(col, tuple)) and col in self:
+                # tuples can be both column keys or list-likes
+                # if they are valid column keys, everything is fine
+                continue
+            elif is_scalar(col) and col not in self:
+                # tuples that are not column keys are considered list-like,
+                # not considered missing
+                missing.append(col)
+            elif (not is_list_like(col) or isinstance(col, set)
```

Comment:
is this isinstance check for set necessary?

Reply:
I didn't know sets are considered list-like. IMO they shouldn't be, because they have no defined order (which is important for an index, especially when `append=True`). Do we really want to support this option of letting users shoot themselves in the foot?
```
>>> s = pd.Series([1, 2, 3])
>>> s.set_index({'a', 'c', 'e'}, append=True)
0  c    1
1  e    2
2  a    3
dtype: int64
```

Reply:
i am not averse to excluding, but i also don't want ad-infinitem special cases here. This code is already too complex. We accept an iterable, so this is a valid input, if the user wants to do it that's there issue; these are not currently excluded.

Reply:
The larger issue IMO is that `is_list_like(<set>)` should be False and not True. It's the only exclusion here, and it makes sense. On many issues, pandas prides itself on enforcing sensible defaults and behaviour. Why would we want to enable something that's broken by design?

I may also add that **this is not something that's currently allowed**, and it shouldn't be IMO